internal/instance/resource_instance.go

```diff
@@ -371,13 +384,73 @@ func (r InstanceResource) ValidateConfig(ctx context.Context, req resource.Valid
 		)
 	}
 
-	if !config.Image.IsNull() && !config.SourceInstance.IsNull() {
+	if !atMostOne(!config.Image.IsNull(), !config.SourceFile.IsNull(), !config.SourceInstance.IsNull()) {
 		resp.Diagnostics.AddError(
 			"Invalid Configuration",
-			"Only image or source_instance can be set.",
+			"At most one of image, source_file and source_instance can be set.",
 		)
 		return
 	}
+
+	if !config.SourceFile.IsNull() {
+		// Instances from source_file are mutually exclusive with a series of other attributes.
+		if !config.Description.IsNull() ||
+			!config.Type.IsNull() ||
+			!config.Ephemeral.IsNull() ||
+			!config.Profiles.IsNull() ||
+			!config.Files.IsNull() ||
+			!config.Config.IsNull() {
+			resp.Diagnostics.AddError(
+				"Invalid Configuration",
+				"Attribute source_file is mutually exclusive with description, type, ephemeral, profiles, file and config.",
+			)
+			return
+		}
+
+		// With `incus import`, a storage pool can be provided optionally.
+		// In order to support the same behavior with source_file,
+		// a single device entry of type `disk` is allowed with a single property
+		// `pool` set to the respective pool name.
+		if len(config.Devices.Elements()) > 0 {
+			if len(config.Devices.Elements()) > 1 {
+				resp.Diagnostics.AddError(
+					"Invalid Configuration",
+					"Only one device entry is supported with source_file.",
+				)
+				return
+			}
+
+			deviceList := make([]common.DeviceModel, 0, 1)
+			diags = config.Devices.ElementsAs(ctx, &deviceList, false)
+			if diags.HasError() {
+				resp.Diagnostics.Append(diags...)
+				return
+			}
+
+			if len(deviceList[0].Properties.Elements()) != 1 {
+				resp.Diagnostics.AddError(
+					"Invalid Configuration",
+					`Exactly one device property named "pool" needs to be provided with source_file.`,
```

Comment:
That's inaccurate, for it to look like a valid Incus device, it needs to be:

  type: disk
  pool: POOLNAME
  path: /

Otherwise it's going to confuse people if we're expected to pass what would otherwise be an invalid device here.